src/lib.rs

```diff
@@ -223,66 +286,145 @@ struct StatementData {
+                io::ErrorKind::ConnectionAborted,
+                "peer terminated connection",
+            )
+        })?;
+        let handshake = commands::client_handshake(&handshake)
+            .map_err(|e| match e {
+                nom::Err::Incomplete(_) => io::Error::new(
+                    io::ErrorKind::UnexpectedEof,
+                    "client sent incomplete handshake",
+                ),
+                nom::Err::Failure(nom_error) | nom::Err::Error(nom_error) => {
+                    if let nom::error::ErrorKind::Eof = nom_error.code {
+                        io::Error::new(
+                            io::ErrorKind::UnexpectedEof,
+                            format!(
+                                "client did not complete handshake; got {:?}",
+                                nom_error.input
+                            ),
+                        )
+                    } else {
+                        io::Error::new(
+                            io::ErrorKind::InvalidData,
+                            format!(
+                                "bad client handshake; got {:?} ({:?})",
+                                nom_error.input, nom_error.code
+                            ),
+                        )
                     }
-                })?
-                .1;
-            self.writer.set_seq(seq + 1);
+                }
+            })?
+            .1;
+
+        self.writer.set_seq(seq + 1);
+
+        let mut auth_response = handshake.auth_response.to_vec();
+        let username = handshake.username.to_vec();
+        let auth_plugin = handshake.auth_plugin.to_vec();
+
+        drop(handshake);
+
+        let user_auth_plugin = self.shim.auth_plugin_for_user(&username);
+        // Start SwitchAuthRequest
+        if user_auth_plugin.as_bytes() != auth_plugin {
+            self.writer.write_all(&[0xfe])?;
+            self.writer.write_all(user_auth_plugin.as_bytes())?;
+            self.writer.write_all(&[0x00])?;
+            self.writer.write_all(&salt)?;
+            self.writer.write_all(&[0x00])?;
+
+            self.writer.flush()?;
+            {
+                let (seq, auth_response_data) = self.reader.next()?.ok_or_else(|| {
+                    io::Error::new(
+                        io::ErrorKind::ConnectionAborted,
+                        "peer terminated connection",
+                    )
+                })?;
+                self.writer.set_seq(seq + 1);
+                auth_response = auth_response_data.to_vec();
+            }
+        }
+
+        if !self
+            .shim
+            .authenticate(user_auth_plugin, &username, &salt, &auth_response)
+        {
+            let err_msg = format!(
+                "Access denied for user '{}'@'{}' (using password: YES)",
+                String::from_utf8_lossy(&username),
+                self.socket_addr.ip()
+            );
+            writers::write_err(
+                ErrorKind::ER_ACCESS_DENIED_ERROR,
```

Comment:
Should we perhaps allow `MysqlShim::authenticate` to return `Result<(), ErrorKind>` so that the implementor can indicate the error type? Or perhaps even `Result<(), (String, ErrorKind)>` so they can also dictate the message?